### ISSUE TITLE
Automatic recovery failures

### ISSUE THREAD

**Issue by MichaelLogutov** [rabbitmq/rabbitmq-dotnet-client#272]
We've a quite a number of problems running multiple console applications (via task scheduler) on virtual machine with a lot of them being rabbit mq consumers using this library.

Basically, our consumers using BasicGet, Ack/Nack in a loop (with a bit of throttle).

The problem is when there is something wrong with rabbitmq server or network the channel disconnected  but fails to automatically recover after that.

Here a sample log of one incedent.

First, we've got this unobserved task exception:

```
28.10.2016 12:34:04 | ERROR | Application | A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread.
System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> RabbitMQ.Client.Exceptions.AlreadyClosedException: Already closed: The AMQP operation was interrupted: AMQP close-reason, initiated by Library, code=0, text="End of stream", classId=0, methodId=0, cause=System.IO.EndOfStreamException: Unable to read beyond the end of the stream.
   at RabbitMQ.Client.Impl.Frame.ReadFrom(NetworkBinaryReader reader)
   at RabbitMQ.Client.Impl.SocketFrameHandler.ReadFrame()
   at RabbitMQ.Client.Framing.Impl.Connection.MainLoopIteration()
   at RabbitMQ.Client.Framing.Impl.Connection.MainLoop()
   at RabbitMQ.Client.Impl.SessionBase.Transmit(Command cmd)
   at RabbitMQ.Client.Impl.ModelBase.ModelRpc(MethodBase method, ContentHeaderBase header, Byte[] body)
   at RabbitMQ.Client.Framing.Impl.Model._Private_ChannelOpen(String outOfBand)
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.CreateNonRecoveringModel()
   at RabbitMQ.Client.Impl.AutorecoveringModel.AutomaticallyRecover(AutorecoveringConnection conn, IConnection connDelegate)
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.RecoverModels()
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.PerformAutomaticRecovery()
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.<>c__DisplayClass84_0.<BeginAutomaticRecovery>b__0()
   at System.Threading.Tasks.Task.Execute()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) RabbitMQ.Client.Exceptions.AlreadyClosedException: Already closed: The AMQP operation was interrupted: AMQP close-reason, initiated by Library, code=0, text="End of stream", classId=0, methodId=0, cause=System.IO.EndOfStreamException: Unable to read beyond the end of the stream.
   at RabbitMQ.Client.Impl.Frame.ReadFrom(NetworkBinaryReader reader)
   at RabbitMQ.Client.Impl.SocketFrameHandler.ReadFrame()
   at RabbitMQ.Client.Framing.Impl.Connection.MainLoopIteration()
   at RabbitMQ.Client.Framing.Impl.Connection.MainLoop()
   at RabbitMQ.Client.Impl.SessionBase.Transmit(Command cmd)
   at RabbitMQ.Client.Impl.ModelBase.ModelRpc(MethodBase method, ContentHeaderBase header, Byte[] body)
   at RabbitMQ.Client.Framing.Impl.Model._Private_ChannelOpen(String outOfBand)
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.CreateNonRecoveringModel()
   at RabbitMQ.Client.Impl.AutorecoveringModel.AutomaticallyRecover(AutorecoveringConnection conn, IConnection connDelegate)
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.RecoverModels()
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.PerformAutomaticRecovery()
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.<>c__DisplayClass84_0.<BeginAutomaticRecovery>b__0()
   at System.Threading.Tasks.Task.Execute()<---
```

In mean time in our trace log:

```
28.10.2016 12:33:30 | DEBUG | Start: ack message 132680
28.10.2016 12:33:49 | DEBUG | Get message from queue
28.10.2016 12:33:50 | DEBUG | Get message from queue
28.10.2016 12:33:55 | DEBUG | Get message from queue
28.10.2016 12:33:57 | DEBUG | Get message from queue
28.10.2016 12:33:59 | DEBUG | Get message from queue
28.10.2016 12:34:01 | DEBUG | Get message from queue
28.10.2016 12:34:02 | DEBUG | Get message from queue
28.10.2016 12:34:04 | DEBUG | Get message from queue
... goes like that for hours ...
```

The first line from this code:

```
public void Ack(IModel channel, ReceivedQueueMessage message)
{
    if (this.configuration.TraceLogging)
        Log.Debug("Start: ack message {0}", message.DeliveryTag);

    channel.BasicAck(message.DeliveryTag, false);

    if (this.configuration.TraceLogging)
        Log.Debug("Complete: ack message {0}", message.DeliveryTag);
}
```

As you can see - it does not returns properly (it actually throws an error as you can see later).
Other lines - a parallel Task calling BasicGet but not receiving anything (it would print "Got one message" otherwise with full json of the message). Nedless to say there are a lot of message in the queue at that time.

So after receiving an error we've got our connection down which is also correlated with the couple more exceptions in error log:

```
28.10.2016 12:34:19 | ERROR | Application | Unable to write data to the transport connection: An established connection was aborted by the software in your host machine.
System.IO.IOException: Unable to write data to the transport connection: An established connection was aborted by the software in your host machine. ---> System.Net.Sockets.SocketException: An established connection was aborted by the software in your host machine
   at System.Net.Sockets.Socket.Send(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   --- End of inner exception stack trace ---
   at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   at System.IO.BufferedStream.Flush()
   at RabbitMQ.Client.Impl.SocketFrameHandler.WriteFrame(Frame frame)
   at RabbitMQ.Client.Impl.Command.TransmitAsSingleFrame(Int32 channelNumber, Connection connection)
   at RabbitMQ.Client.Impl.SessionBase.Transmit(Command cmd)
   at RabbitMQ.Client.Impl.ModelBase.ModelSend(MethodBase method, ContentHeaderBase header, Byte[] body)
   at RabbitMQ.Client.Impl.RecoveryAwareModel.BasicAck(UInt64 deliveryTag, Boolean multiple)
   at OurCode.RabbitClientService.Ack(IModel channel, ReceivedQueueMessage message)
   at OurCode.Ack(IModel channel, ReceivedQueueMessage message)


28.10.2016 12:35:22 | ERROR | Application | Already closed: The AMQP operation was interrupted: AMQP close-reason, initiated by Library, code=0, text="End of stream", classId=0, methodId=0, cause=System.IO.IOException: Unable to write data to the transport connection: An established connection was aborted by the software in your host machine. ---> System.Net.Sockets.SocketException: An established connection was aborted by the software in your host machine
   at System.Net.Sockets.Socket.Send(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   --- End of inner exception stack trace ---
   at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   at System.IO.BufferedStream.Flush()
   at RabbitMQ.Client.Impl.SocketFrameHandler.WriteFrame(Frame frame)
   at RabbitMQ.Client.Framing.Impl.Connection.HeartbeatWriteTimerCallback(Object state)
RabbitMQ.Client.Exceptions.AlreadyClosedException: Already closed: The AMQP operation was interrupted: AMQP close-reason, initiated by Library, code=0, text="End of stream", classId=0, methodId=0, cause=System.IO.IOException: Unable to write data to the transport connection: An established connection was aborted by the software in your host machine. ---> System.Net.Sockets.SocketException: An established connection was aborted by the software in your host machine
   at System.Net.Sockets.Socket.Send(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   --- End of inner exception stack trace ---
   at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   at System.IO.BufferedStream.Flush()
   at RabbitMQ.Client.Impl.SocketFrameHandler.WriteFrame(Frame frame)
   at RabbitMQ.Client.Framing.Impl.Connection.HeartbeatWriteTimerCallback(Object state)
   at RabbitMQ.Client.Impl.SessionBase.Transmit(Command cmd)
   at RabbitMQ.Client.Impl.ModelBase.ModelSend(MethodBase method, ContentHeaderBase header, Byte[] body)
   at RabbitMQ.Client.Impl.RecoveryAwareModel.BasicNack(UInt64 deliveryTag, Boolean multiple, Boolean requeue)
   at OurCode.RabbitClientService.Nack(IModel channel, ReceivedQueueMessage message)
   at OurCode.QueueService.HandleMessageProcessinigError(ProcessQueueMessagesParams p, IModel channel, ReceivedQueueMessage message, Exception ex)
```

After that application can forever cal BasicGet which does not signals in any way that connection is down and can not be recovered - it's just returns null, as if the queue was empty! Running wireshark in that time confirmed that there were no BasicGet commands send for this queue at that time - so library didn't actually called the server - it's like it tries to use broken connection, swallows the exception and returns null as if nothing happened.

We've tried to set UseBackgroundThreadsForIO to false - does nothing (suprisingly we're still getting unobserved task exception - which is not what I was expected by switching of the settings with that name).

We even tried AutomaticRecoveryEnabled and TopologyRecoveryEnabled both to false in hope to receive proper exception in BasicGet on channel down - nothing. It silently returns null.

I don't understand how we should use this library to reliably reconnect in case of channel failures. BasicGet does not throws exception for us to maybe just throw away IConnection and recreate it. And in the same time automatic recovery does nothing for more than an half an hour, while there are other consumers on the same machine that were consuming messages from the same server (but from different queues) successfully at the same time.

I really whish there were some kind of dead simple HTTP-like Get/Publish/Ack/Nack API without that fancy background threads, locks and autorecoveries if possible - at least we could try to properly react to channel failures and clearly understand what happening.

We're using https://www.nuget.org/packages/RabbitMQ.Client/4.1.1-rc1

**Comment by michaelklishin** [comment]
You are mixing together two different things : BasicGet error reporting and recovery. I don't see any evidence that the latter didn't work as it was supposed to. Therefore the issue title is misleading.

**Comment by michaelklishin** [comment]
Also, consider using a consumer instead of BasicGet. BasicGet suffers from the same fundamental issues with recovery that QueueingBasicConsumer does and for that we do not claim to support automatic recovery because there is no way to do it without application developer's effort in general (with other consumers it is fairly easy to reason about what a successful recovery should look like).

This and my earlier comment on what this issue really seems to be about leads me to believe this is mailing list material and not an actionable item at this point.

**Comment by MichaelLogutov** [comment]
So QueueingBasicConsumer does have some code for proper handling those cases? Last time I checked it was quite simple thing with blocking queue, if I'm not mistaken.

**Comment by michaelklishin** [comment]
It doesn't. It is not supported by automatic recovery and these days there are no reasons to use it at all.

I can explain why it was once necessary once I am near a physical keyboard.

**Comment by MichaelLogutov** [comment]
Ok, so whats recommended way to work with this library? We were using BasicGet and manual Ack/Nack because we've got TPL for processing messages and didn't willing to introduce more complexity to the pipeline.

Thanks for answers, I'll wait till you get to more usable keyboard :)

**Comment by MichaelLogutov** [comment]
Hello. Any news on that? You've said I should use a consumer, but not QueueingBasicConsumer - which one then? And what does that consumer have to overcome those problems?

**Comment by kjnilsson** [comment]
`EventingBasicConsumer` is the replacement for `QueuingBasicConsumer`. `QueuingBasicConsumer` has a variety of issues and is generally fairly hard to reason about. `EventingBasicConsumer` should be a bit easier to work with.

I wasn't able to reproduce the error you reported in https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/260 - Is this related or a separate issue?

**Comment by MichaelLogutov** [comment]
They could be related. We have heavy loaded VM and this can be the reason there are so many edge cases with network stability.

About EventingBasicConsumer - as I've said we've got TPL dataflow and we need to incorporate our ack/nack and parallel message processing in EventingBasicConsumer (if it even possible).

**Comment by kjnilsson** [comment]
Can't you just dispatch arriving messages into your TPL dataflow components? It has been a while since I last looked at TPL dataflow so am not sure what the constraints are.

**Comment by MichaelLogutov** [comment]
Will do. Any thought of what's so special about EventingBasicConsumer that could help with channel stability?

**Comment by michaelklishin** [comment]
Questions belong to rabbitmq-users.

**Comment by michaelklishin** [comment]
Here's [why QueueingConsumer was introduced and is no longer relevant](https://github.com/rabbitmq/rabbitmq-java-client/commit/7b16a33abf6e4c3c5c799bcdc869b653b14b3542#commitcomment-19719162). That comment is for the Java client but everything said there is applicable to this client as well.

**Comment by MichaelLogutov** [comment]
Thanks. We've changed our library to eventing consumer (btw there is still "experimental" comment on the eventing consumer class description) - our consumers still disconnecting but now they are able to connect back. Tho it seems that thay wait like 10 minutes or so. But now we suspect that there maybe some port exhaustion or something other happening.

**Comment by michaelklishin** [comment]
@MichaelLogutov take a look at [the heartbeats guide](http://www.rabbitmq.com/heartbeats.html).

**Comment by MichaelLogutov** [comment]
Previously we've set all timeouts (connection, heartbeat, etc) to lower values than defaults in attempt for consumers to faster reconnect in case of network failures. Settings those config values back to original (except heartbeat - I've set it to 10 sec and now increased to 20 sec) seems to improve overall system health. Now we see almost no connection dropping. It seems that somehow lower those values produced a lot of reconnections which had put a lot of pressure to system socket pool, producing strange errors all over the stack (like DNS issues). It's a bit early to judge if we've managed to solve this issue but situation is a lot more healthy now. Thanks again.

**Comment by michaelklishin** [comment]
There is a section on heartbeat false positives in the docs. I'm sorry but reconnections causing DNS resolution issues doesn't make much sense.

**Comment by MichaelLogutov** [comment]
Yeah, we though that too. Until one day an application running on local computer eated up resources such way, that machine was working fine, but there were no internet connection for any program. And there were low number of opened connections, low number of TCP/UDP channels etc. - all was working fine but chrome and other programs fail to connect to any site. Running some other programs got us "socket buffer  overflow" exception. Shutting down the program immediately restored internet access for all programs.

**Comment by michaelklishin** [comment]
Socket exhaustion is a rare meaningful explanation. My point was that it can also work the other way around: resolution timeouts can lead to connections building up.
